webtransport/README.md

```diff
@@ -0,0 +1,62 @@
+
+Servers need to take care to regularly renew their certificate. In the following, the RECOMMENDED logic for rolling certificates is described. At first boot of the node, it creates a self-signed certificate with a validity of 14 days, and publishes a multiaddr containing the hash of that certificate. After 10 days, the node prepares the next certificate, setting the `NotBefore` date of that certificate to the expiration date (or shortly before that) of the first certificate, and an expiration of 14 days after that. The node continues using the old certificate until the expiry date, but it already advertises a multiaddr containing both certificate hashes. This way, clients will be able to connect to the node, even if they cache the multiaddr for multiple days.
+
+## Addressing
+
+Webtransport multiaddresses are composed of a QUIC multiaddr, followed by `/webtransport` and a list of multihashes of the certificates that the server uses.
+Examples:
+* `/ip4/1.2.3.4/udp/443/quic/webtransport/certhash/<hash1>`
+* `/ip6/fe80::1ff:fe23:4567:890a/udp/1234/quic/webtransport/certhash/<hash1>/certhash/<hash2>/certhash/<hash3>`
+
+## WebTransport HTTP endpoint
+
+WebTransport needs a HTTPS URL to establish a WebTransport session, e.g. `https://example.com/webtransport`. As multiaddrs don't allow the encoding of URLs, this spec standardizes the endpoint. The HTTP endpoint of a libp2p WebTransport servers MUST be located at `/.well-known/libp2p-webtransport`.
+
+To allow future evolution of the way we run the libp2p handshake over WebTransport, we use a URL parameter. The handshake described in this document MUST be signaled by setting the `type` URL parameter to `noise`.
+
+Example: The WebTransport URL of a WebTransport server advertising `/ip4/1.2.3.4/udp/1443/quic/webtransport/` would be `https://1.2.3.4:1443/.well-known/libp2p-webtransport?type=noise`.
+
+## Security Handshake
+
+Unfortunately, the self-signed certificate doesn't allow the nodes to authenticate each others' peer IDs. It is therefore necessary to run an additional libp2p handshake on a newly established WebTransport connection.
+The first stream that the client opens on a new WebTransport session is used to perform a libp2p handshake using Noise (https://github.com/libp2p/specs/tree/master/noise). The client SHOULD start the handshake right after sending the CONNECT request, without waiting for the server's response.
+
+In order to verify that the end-to-end encryption of the connection, the peers need to establish that no MITM intercepted the connection. To do so, the client MUST include the certificate hash that was used to establish the connection as payload of the first Noise message (the `e` message). This payload is not encrypted, but the Noise handshake provides integrity protection.
```

Comment:
Alternatively, we could have the server send the hashes of all its certificates in the second handshake message. This might mean that we'd transfer one more certificate hash in some cases, but the overhead should be negligible.
The two advantages are:
1. This message would be encrypted.
2. It would shift the responsibility of verifying the certificate hashes onto the client.

@MarcoPolo @julian88110 @mxinden Thoughts?

Reply:
This handshake is encrypted by the peer ids. The client should see the superset of the certhashes it used to start the connection sent by the expected peer id. I think this is the same as before, except of course the client gets to be sure the check was done. So I think this is a net win.

Reply:
> This message would be encrypted.

:+1: for choosing this new approach then.

> It would shift the responsibility of verifying the certificate hashes onto the client.

I can not think of a reason for either being beneficial over the other. I.e. for server or client verification to be better. Thus this does not effect my agreement with your proposal. Is there any reason?

> I think this is the same as before, except of course the client gets to be sure the check was done.

I read this as there being a reason. @MarcoPolo would you mind expanding if so?

Reply:
> I read this as there being a reason. @MarcoPolo would you mind expanding if so?

Assume server keys aren't leaked (then all bets are off).

If the server is responsible for the verification then the client has to trust that the server (or the server's implementation) does the correct check. If the client does the verification, it doesn't have to trust that the server did this check it can do it itself. @marten-seemann mentioned this in his comment: https://github.com/libp2p/specs/pull/404#discussion_r846792263. The flip side of course, is that the server now trusts the client to do the check to avoid mitm, but this is how it works everywhere. In web2, servers can't force the browser to actually check the signature, but they do.

The point of the certificate hash is so that the client can validate that the server it's talking to is indeed the server it expects. The key thing here is that the _client_ cares about this. The client has no reason to believe that the server cares about this. In classic webtransport this is validated on the client side as well (this should be reason enough to do the same thing here), granted this needs to happen because clients don't know the server's publickey/peerid. 

Which brings up another thing you enable with client-side checking. You could say you want to dial any peer id that passes a given certhash check. So if you got a certhash over a trusted channel, you could use it to authenticate the connection without knowing the peer id (this is the classic webtransport case).

Reply:
Text change suggested in #455.